analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.Arithmetic.cs

```diff
@@ -0,0 +1,133 @@
+
+internal sealed partial class Binary : BranchingProcessor<IBinaryOperationWrapper>
+{
+    private static NumberConstraint Calculate(BinaryOperatorKind kind, NumberConstraint left, NumberConstraint right) => kind switch
+    {
+        BinaryOperatorKind.Add => NumberConstraint.From(left.Min + right.Min, left.Max + right.Max),
+        BinaryOperatorKind.Subtract => NumberConstraint.From(left.Min - right.Max, left.Max - right.Min),
+        BinaryOperatorKind.Multiply => CalculateMultiply(left, right),
+        BinaryOperatorKind.And when left.IsSingleValue && right.IsSingleValue => NumberConstraint.From(left.Min.Value & right.Min.Value),
+        BinaryOperatorKind.And => CalculateAnd(left, right),
+        _ => null
+    };
+
+    private static NumberConstraint CalculateMultiply(NumberConstraint left, NumberConstraint right)
+    {
+        var products = new[] { left.Min * right.Min, left.Min * right.Max, left.Max * right.Min, left.Max * right.Max };
+        var min = products.Min();
+        var max = products.Max();
+
+        if ((left.Min is null && right.CanBePositive) || (right.Min is null && left.CanBePositive)
+            || (left.Max is null && right.CanBeNegative) || (right.Max is null && left.CanBeNegative))
+        {
+            min = null;
+        }
+        if ((left.Min is null && right.CanBeNegative) || (right.Min is null && left.CanBeNegative)
+            || (left.Max is null && right.CanBePositive) || (right.Max is null && left.CanBePositive))
+        {
+            max = null;
+        }
+
+        return NumberConstraint.From(min, max);
+    }
+
+    private static NumberConstraint CalculateAnd(NumberConstraint left, NumberConstraint right)
+    {
+        // The result can only be negative if both operands are negative => The result must be >= 0 unless both ranges include negative numbers.
+        BigInteger? min = 0;
+        if (left.CanBeNegative && right.CanBeNegative)
+        {
+            min = left.Min.HasValue && right.Min.HasValue
+                ? NegativeMagnitude(BigInteger.Min(left.Min.Value, right.Min.Value))
+                : null;
+        }
+        return NumberConstraint.From(min, CalculateAndMax(left, right));
+    }
+
+    private static BigInteger? CalculateAndMax(NumberConstraint left, NumberConstraint right)
+    {
+        if ((left.IsNegative && right.IsNegative)
+            || (left.IsPositive && right.IsPositive))
+        {
+            return SmallestMaximum(left, right);
+        }
+        // -1 == 0b11111111 => a & -1 == a =>
+        // If one operand can be -1 and one is positive, the result will be positive, but not bigger than the positive operand.
+        else if ((left.IsNegative && right.CanBePositive)                          // left has no positive values => max must come from right
+            || (left.CanBePositive && left.CanBeNegative && right.IsPositive))     // right cannot be -1 => max must come from right
```

Comment:
The first `if` above already ruled out that `left.IsPositive`. So we're sure that `left.CanBeNegative` => we don't need to ask.

At the same time, we do we ask if `left.CanBePositive`? If I remove it, no UT breaks. 

```suggestion
        else if ((left.IsNegative && right.CanBePositive)                          // left has no positive values => max must come from right
            || (right.IsPositive))     // right cannot be -1 => max must come from right
```